web/src/components/UserProfile.tsx

```diff
@@ -0,0 +1,113 @@
+import React, { useState } from "react";
+
+import "./UserProfile.css";
+import {
+  Emoji,
+  EmojiPicker,
+  EmojiDataType,
+  isEmojiDataType,
+} from "../utils/emoji";
+
+const TextField = React.memo<{
+  initialText: string;
+  onUpdate: (text: string) => void;
+  buttonLabel?: string;
+  placeholder?: string;
+  clearOnUpdate?: boolean;
+}>(({ initialText, onUpdate, buttonLabel, placeholder, clearOnUpdate }) => {
+  const [text, setText] = useState(initialText);
+  const onSubmit = (event: React.FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    if (text) {
+      onUpdate(text);
+      if (clearOnUpdate) {
+        setText("");
+      }
+    }
+  };
+
+  return (
+    <form onSubmit={onSubmit}>
+      <input
+        value={text}
+        onChange={(e) => setText(e.target.value)}
+        placeholder={placeholder}
+      />
+      {buttonLabel && (
+        <button type="submit" disabled={!text}>
+          {buttonLabel}
+        </button>
+      )}
+    </form>
+  );
+});
+
+const UserProfile: React.FC<{
+  initialNickname: string;
+  emoji: EmojiDataType | null;
+  onUpdateNickname: (e: string) => void;
+  onUpdateStatusMesg: (e: string) => void;
+  onUpdateEmoji: (e: EmojiDataType | null) => void;
+}> = ({
```

Comment:
I think I will make other components consistent with this pattern without `type Props = { ... }`.

Reply:
fixed in 424d484